### PR TITLE
Add a receiver for Amazon SNS

### LINKED ISSUE #172: Add Amazon SNS as receiver
As noted [here](https://github.com/grafana/alerting/issues/20) Amazon SNS is missing as an alert receiver in Grafana, while supported in Mimir

This package should support Amazon SNS as a receiver to enable Grafana to send alert notifications to SNS for Grafana-managed alerts. Amazon Managed Grafana supports this today, but it is not present in the open source Grafana. 

I can contribute this.

### PR DESCRIPTION
**Pull request by robbierolin**
This PR adds a receiver that will send notifications to Amazon SNS. Fixes https://github.com/grafana/alerting/issues/172

I have a corresponding change in the grafana code locally to add this receiver to the available channels that can be selected for contact points. I am guessing the process is to PR that one after this change gets merged.

Here are some example screenshots

Screenshot 2024-03-20 at 5 35 49 PM

I tested by subscribing my email to the sns topic. And verifying that I receive the emails when I send test requests.

Screenshot 2024-03-20 at 5 36 51 PM

Let me know if there's any other changes I should add or testing I should show.

### CODE REVIEW

File: receivers/sns/config.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		Topic         string `json:"topic,omitempty" yaml:"topic,omitempty"`
+		AuthProvider  string `json:"authProvider,omitempty" yaml:"authProvider,omitempty"`
+		Subject       string `json:"subject,omitempty" yaml:"subject,omitempty"`
+		Body          string `json:"body,omitempty" yaml:"body,omitempty"`
+		MessageFormat string `json:"messageFormat,omitempty" yaml:"messageFormat,omitempty"`
+		Credentials   string `json:"credentials,omitempty" yaml:"credentials,omitempty"`
+		AccessKey     string `json:"accessKey,omitempty" yaml:"accessKey,omitempty"`
+		SecretKey     string `json:"secretKey,omitempty" yaml:"secretKey,omitempty"`
+		AssumeRoleARN string `json:"assumeRoleARN,omitempty" yaml:"assumeRoleARN,omitempty"`
+		ExternalID    string `json:"externalId,omitempty" yaml:"externalId,omitempty"`
```
Comment (yuri-tceretian) [inline receivers/sns/config.go]: First, thank you so much for opening the PR! 

I have not reviewed it yet but the first thing that catches my eye is the config. We're trying to maintain as little drift as possible with integration in the upstream. Please make sure that json\yaml fields match the upstream integration 
https://github.com/prometheus/alertmanager/blob/f82574a3763c21ca9bf2ada7baf69bd09e01d977/config/notifiers.go#L731-L744

Comment (robbierolin) [reply receivers/sns/config.go]: Thanks for sharing, I will look at that. I think I will need to use this [SubformOptions](https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/notifier/channels_config/plugin.go#L27) to provide the complex types like [sigv4.SigV4Config](https://github.com/prometheus/common/blob/main/sigv4/sigv4_config.go#L25)?

Comment (yuri-tceretian) [reply receivers/sns/config.go]: Yes, complex types can be presented via subform options. You can see https://github.com/grafana/grafana/blob/cf601fab0906f1d811f1aa2fe18a67bc5421e3db/pkg/services/ngalert/notifier/channels_config/available_channels.go#L1312-L1338 as an example

Comment (robbierolin) [reply receivers/sns/config.go]: Ok I updated the config to match the upstream and pulled in some of the functionality from there too, like message deduplication and truncation.

Also I noticed that the Grafana channel config UIs doesn't seem to support ShowWhen or secure fields for subforms. I can try to address that in my PR to grafana.

**Comment by JohnnyQQQQ** [comment]
@robbierolin could you rebase your PR please and @yuri-tceretian could you give it a final review?

**Comment by robbierolin** [comment]
Yup, should be good now.

File: receivers/sns/config.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	at := awsds.AuthTypeDefault
+	if settings.Sigv4.Profile != "" {
+		at = awsds.AuthTypeSharedCreds
+	} else if settings.Sigv4.AccessKey != "" && settings.Sigv4.SecretKey != "" {
+		settings.Sigv4.AccessKey = decryptFn("accessKey", settings.Sigv4.AccessKey)
+		settings.Sigv4.SecretKey = decryptFn("secretKey", settings.Sigv4.SecretKey)
+		at = awsds.AuthTypeKeys
+	}
```
Comment (yuri-tceretian) [inline receivers/sns/config.go:88]: Can we add a validation here to check for situations when either of keys is not specified.  Currently, it silently falls back to default authentication type

File: receivers/sns/config_test.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			expectedAuthType: awsds.AuthTypeDefault.String(),
+		},
+		{
+			name:             "Full config gives no errors",
+			settings:         FullValidConfigForTesting,
+			expectedSubject:  "subject",
+			expectedMessage:  "message",
+			expectedAuthType: awsds.AuthTypeSharedCreds.String(),
+		},
+	}
```
Comment (yuri-tceretian) [inline receivers/sns/config_test.go:101]: Can you add tests for invalid configurations like
```api
settings: `{
				"topic_arn": "arn:aws:sns:region:0123456789:SNSTopicName",
				"sigv4": {
					"secret_key": "secret-key"
				}
			}`,
```

and 
```api
settings: `{
"topic_arn": "arn:aws:sns:region:0123456789:SNSTopicName",
"sigv4": {
"access-key": "access-key"
}
}`,
```

File: receivers/sns/config.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		settings.Subject = templates.DefaultMessageTitleEmbed
+	}
+	if settings.Message == "" {
+		settings.Message = templates.DefaultMessageEmbed
+	}
+	if settings.APIUrl == "" {
+		settings.APIUrl = fmt.Sprintf("https://sns.%s.amazonaws.com", settings.Sigv4.Region)
+	}
+
+	at := awsds.AuthTypeDefault
```
Comment (yuri-tceretian) [inline receivers/sns/config.go:78]: Do I understand correctly that in the case when auth type is "default", the credentials will be taken from a file `~/.aws/credentials` or env vars?

File: receivers/sns/sns.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+func validateAndTruncateMessage(message string, maxMessageSizeInBytes int) (string, bool, error) {
+	if !utf8.ValidString(message) {
+		return "", false, fmt.Errorf("non utf8 encoded message string")
+	}
+	if len(message) <= maxMessageSizeInBytes {
+		return message, false, nil
+	}
+	// If the message is larger than our specified size we have to truncate.
+	truncated := make([]byte, maxMessageSizeInBytes)
```
Comment (yuri-tceretian) [inline receivers/sns/sns.go:143]: We have a standard way for truncating messages `notify.TruncateInBytes`. However, this matches the logic in the vanilla Alertmanager. So, I think this is fine.

File: receivers/sns/sns_test.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/common/model"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/alerting/logging"
+	"github.com/grafana/alerting/receivers"
+	"github.com/grafana/alerting/templates"
+)
+
+func TestCreatePublishInput(t *testing.T) {
```
Comment (yuri-tceretian) [inline receivers/sns/sns_test.go:18]: This needs some more tests for truncation of messages

**Review by yuri-tceretian — COMMENTED** [review]
It looks like the logic pretty much matches already merged in vanilla Alertmanager. The code LGTM but I think it needs some more tests.

**Comment by yuri-tceretian** [comment]
CI fails due to linter
```
golangci-lint run

receivers/sns/sns.go:6: File is not `goimports`-ed (goimports)

	"github.com/prometheus/alertmanager/notify"

receivers/sns/config_test.go:4: File is not `goimports`-ed (goimports)

	"encoding/json"
```

File: receivers/sns/config.go (hunk 4 above)
Comment (robbierolin) [reply receivers/sns/config.go:78]: That setting comes from here - https://github.com/grafana/grafana-aws-sdk/blob/main/pkg/awsds/settings.go#L16
The behaviour should be to use the default credential provider chain which looks for creds in multiple places. It's explained more here - https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials

File: receivers/sns/config.go (hunk 2 above)
Comment (robbierolin) [reply receivers/sns/config.go:88]: Yes, will do.

File: receivers/sns/config_test.go (hunk 3 above)
Comment (robbierolin) [reply receivers/sns/config_test.go:101]: added

File: receivers/sns/sns_test.go (hunk 6 above)
Comment (robbierolin) [reply receivers/sns/sns_test.go:18]: added

**Review by yuri-tceretian — APPROVED** [review]
LGTM. Thanks for the contribution!
Once this is merged, we need to update this module in Grafana and introduce a new configuration in https://github.com/grafana/grafana/blob/fdaa091a4d1874e78040a46d2c0d15d29d0064ae/pkg/services/ngalert/notifier/channels_config/available_channels.go#L14
 and update Terraform provider https://github.com/grafana/terraform-provider-grafana/blob/aaa20a20f309172c0b88071f80b9e9e5e52c75ae/internal/resources/grafana/resource_alerting_contact_point_notifiers.go#L2306-L2305
I would appreciate your help with this

**Pull request merged**